### ISSUE TITLE
在执行ppo_tuning.sh的时候出现Loss is nan错误

### ISSUE THREAD

**Issue by xlwreally** [HITSZ-HLT/T2S-Augmentation#2]
感谢作者提供的分享，我参照作者在14res平台上运行的相关命令操作，我使用的是Pytorch 2.1.2，仅针对模型路径做了调整，即将parrot_fluency_model和t5-large的存储位置改至本地，并为了适应4090显卡的硬件条件，减小了batch_size大小。

然而，在实际运行过程中，我在执行train_fluency_discriminator.sh和train_alignment_discriminator.sh脚本时遇到了问题，具体表现为训练过程中的loss值未能实现有效收敛（如图所示）。

最终在执行ppo_tuning.sh脚本阶段，系统报出“Loss is nan”的错误。

![QQ截图20240322093002](https://github.com/HITSZ-HLT/T2S-Augmentation/assets/42504909/1847598c-43f5-46aa-a1bd-c7dc0960500e)

![QQ截图20240322133446](https://github.com/HITSZ-HLT/T2S-Augmentation/assets/42504909/e6063986-7c39-42a2-b7f1-831ec76de9a4)
![QQ截图20240323155522](https://github.com/HITSZ-HLT/T2S-Augmentation/assets/42504909/5b3a6bc7-3469-4598-ac2d-97f08ca9867f)
非常感谢您花时间阅读这个问题，如果您能提供一些指导或建议，我将不胜感激。如果需要更多细节或信息，请随时告知。

**Comment by 1140310118** [comment]
根据我的经验，PPO算法是在运行的时候不是很稳定，容易出现像上面的loss is nan的问题，做超参调整或许可以避免这个问题。在这里，train_alignment_discriminator.sh没有收敛，肯定会极大地影响PPO算法的运行的。train_alignment_discriminator.sh中减小了batch size，按理来说应该重新调一调学习率试试。fluency_discriminator对最后性能的影响不大，可以去掉。

**Comment by xlwreally** [comment]
在更换 RTX A6000并重新配置环境后程序恢复了正常，其中最后两行命令会报错，我修正了命令之后程序得以恢复正常运行：
```
bash/data_synthesis.sh -c 0 -b 14res_100k -g ./output/generation_ppo/model/b=ppo -e ./output/extraction/model/dataset=origin/14res,b=extractor,seed=42 -a ./output/alignment_model/model/b=alignment_model -f ./output/fluency_model/model/b=fluency_model -n 100000 -d data/origin_syn/14res -r output/extraction/pseudo_labeled_syn/yelp2023.json
python data_filtering.py --origin_data_dir data/origin/14res --augmented_data_dir ./output/augmentation/14res_100k_42.json/augmented_train.json --output_dir ./output/augmentation_filtered/14res_5k_42 --k 5000
```
最后在python data_filtering.py 代码报错：
![QQ截图20240327102314](https://github.com/HITSZ-HLT/T2S-Augmentation/assets/42504909/61afc804-8de3-4b1d-85c8-7c81f44280a6)
经过检查我发现在augmented train.json中只有一条数据，这是导致报错的原因。
![QQ截图20240327102348](https://github.com/HITSZ-HLT/T2S-Augmentation/assets/42504909/e87410de-8e6a-49b8-bd97-f12f182079b6)

**Comment by 1140310118** [comment]
你好！非常抱歉，readme中的命令的确有误。非常感谢你指出了这个问题，并提供了更正后的命令！另外，data_synthesis.py的问题现在已经被修复。这个工作整体上还是比较复杂，跑起来比较繁琐，再次感谢你不辞辛劳的follow！

**Comment by xlwreally** [comment]
您好，在data_synthesis.py中依然有缩进问题，我帮你改正，现在data_synthesis.py可以在我的环境中正常运行并输出结果：
 ```python
import os
import random
import math
import argparse
from tqdm import tqdm
import torch
import pytorch_lightning as pl 
pl.seed_everything(42)


from pytorch_lightning.callbacks import BasePredictionWriter
from transformers import T5ForConditionalGeneration, AutoTokenizer
from torch.utils.data import DataLoader
from collections import defaultdict

from model.t5_for_classification import T5ForClassification
from utils import save_json, load_json
from utils.triplet import make_triplets_seq



def get_synseq_len(syn_seq):
    return syn_seq.count(' ')+1


def get_triplet_num_aspect_num_opinion_num(triplets_or_triplets_seq):
    aspects  = []
    opinions = []

    if type(triplets_or_triplets_seq) is str:
        triplet_num = triplets_or_triplets_seq.count(';') + 1
        for triplet_seq in triplets_or_triplets_seq.split(';'):
            aspect = triplet_seq.split('|')[0].strip()
            aspects.append(aspect)
            
            opinion = triplet_seq.split('|')[1].strip()
            opinions.append(opinion)
    else:
        triplet_num = len(triplets_or_triplets_seq)
        for triplet in triplets_or_triplets_seq:
            aspect = triplet['aspect'][-1]
            aspects.append(aspect)
            
            opinion = triplet['opinion'][-1]
            opinions.append(opinion)

    return triplet_num, len(aspects), len(opinions)



class DataModule(pl.LightningDataModule):
    def __init__(self, 
                 generator_path: str='',
                 max_seq_length: int = -1,
                 eval_batch_size: int = 32,
                 data_dir: str = '',
                 reference_data_dir: str = '',
                 seed: int = 42,
                 num_augment_example: int = 12660,
                 sample_triplet_from_where: str = 'reference'):

        super().__init__()
        self.generator_path     = generator_path
        self.max_seq_length     = max_seq_length
        self.eval_batch_size    = eval_batch_size
        self.data_dir           = data_dir
        self.reference_data_dir = reference_data_dir
        self.seed               = seed
        self.num_augment_example= num_augment_example
        self.sample_triplet_from_where = sample_triplet_from_where

        self.tokenizer = AutoTokenizer.from_pretrained(generator_path, use_fast=True)

    def load_dataset(self):
        train_file_name = os.path.join(self.data_dir, 'train.json')
        self.train_examples = load_json(train_file_name)

        self.reference_examples = [
            example for example in load_json(self.reference_data_dir)
            if example['min_con'] > 0.75 and example['avg_con'] > 0.95
        ]

        self.index_example_by_tao(self.reference_examples)
        self.augmented_examples = self.make_example(num_example=self.num_augment_example)


        print('-----------data statistic-------------')
        print('Train',     len(self.train_examples))
        print('Reference', len(self.reference_examples))
        print('Augmented', len(self.augmented_examples))

        print('Train')
        self.stat_triplet_num(self.train_examples)
        print('Reference')
        self.stat_triplet_num(self.reference_examples)
        print('Augmented')
        self.stat_triplet_num(self.augmented_examples)

    def stat_triplet_num(self, examples):
        counter = defaultdict(int)
        for example in examples:
            triplets_seq = (example['triplets_seq'] if 'triplets_seq' in example
                            else make_triplets_seq(example['triplets']))
            tao = get_triplet_num_aspect_num_opinion_num(triplets_seq)
            triplet_num = tao[0]
            counter[triplet_num] += 1

        for triplet_num in sorted(counter.keys()):
            print(triplet_num, ':', counter[triplet_num])

    def index_example_by_tao(self, examples):
        self.example_indexed_by_tao = defaultdict(list)

        for example in examples:
            triplets_seq = (example['triplets_seq'] if 'triplets_seq' in example
                            else make_triplets_seq(example['triplets']))
            tao = get_triplet_num_aspect_num_opinion_num(triplets_seq)

            self.example_indexed_by_tao[tao].append({
                'syn_seq': example['syn_seq'], 
                'triplets_seq': triplets_seq, 
                'sentence': example['sentence']
            })

        print('-----------syn statistic-------------')
        for triplet_num in range(1, 7+1):
            examples = []
            for tao in self.example_indexed_by_tao:
                if tao[0] == triplet_num:
                    _examples = self.example_indexed_by_tao[tao]
                    examples.extend(_examples)

            print(triplet_num, len(examples), 
                  sum([get_synseq_len(example['syn_seq']) for example in examples]) / len(examples))

    def make_example(self, num_example):
        new_examples = []
        num_train_examples = len(self.train_examples)

        for i in range(num_example):
            example = self.train_examples[i%num_train_examples]
            triplets_seq = make_triplets_seq(example['triplets'])
            tao = get_triplet_num_aspect_num_opinion_num(triplets_seq)

            if self.sample_triplet_from_where == 'train':
                m_triplets_seq = make_triplets_seq
                triplet_sentence = example['sentence']

            elif self.sample_triplet_from_where == 'reference':
                m_triplets_seq, triplet_sentence = self.sample_triplets_by_tao(tao)
                tao = get_triplet_num_aspect_num_opinion_num(m_triplets_seq)

            syn_seq, syn_sentence = self.sample_syn_by_tao(tao)
            new_example = {
                'ID': f"{i}",
                'triplets_seq': m_triplets_seq,
                'triplet_sentence': triplet_sentence,
                'syn_seq': syn_seq,
                'syn_sentence': syn_sentence,
            }

            new_examples.append(new_example)

        return new_examples

    def sample_syn_by_tao(self, tao):
        examples = self.get_examples_by_tao(tao)
        random_example = random.choice(examples)
        return random_example['syn_seq'], random_example['sentence']

    def sample_triplets_by_tao(self, tao):
        examples = self.get_examples_by_tao(tao)
        random_example = random.choice(examples)
        return random_example['triplets_seq'], random_example['sentence']
    
    def get_examples_by_tao(self, tao):
        if tao in self.example_indexed_by_tao:
            examples = self.example_indexed_by_tao[tao]
        else:
            most_similarity_tao = min([(abs(tao[0]-t), abs(tao[1]-a), abs(tao[2]-o), (t,a,o)) for t,a,o in self.example_indexed_by_tao.keys()])[-1]
            examples = self.example_indexed_by_tao[most_similarity_tao]
        return examples

    def predict_dataloader(self):
        dataloader = DataLoader(
            dataset=self.augmented_examples,
            batch_size=self.eval_batch_size,
            shuffle=False,
            pin_memory=True,
            prefetch_factor=8,
            num_workers=16,
            collate_fn=DataCollator(
                tokenizer=self.tokenizer, 
                max_seq_length=self.max_seq_length,
            )
        )
        return dataloader



def tok(tokenizer, text, max_seq_length):
    kwargs = {
        'text': text,
        'return_tensors': 'pt'
    }

    if max_seq_length in (-1, 'longest'):
        kwargs['padding'] = True

    else:
        kwargs['padding'] = True
        kwargs['truncation'] = True
        kwargs['max_length'] = max_seq_length

    batch_encodings = tokenizer(**kwargs)
    return batch_encodings




class DataCollator:
    def __init__(self, tokenizer, max_seq_length):
        self.tokenizer = tokenizer
        self.max_seq_length = max_seq_length

    def tok(self, text):
        return tok(self.tokenizer, text, self.max_seq_length)

    def __call__(self, examples):
        conditions = []

        for example in examples:
            syn_seq = example['syn_seq'].lower()
            triplets_seq = example['triplets_seq']
            condition = triplets_seq + ' ; ' + syn_seq
            conditions.append(condition)

        batch_encodings = self.tok(conditions)
        input_ids = batch_encodings['input_ids']
        attention_mask = batch_encodings['attention_mask']

        return {
            'input_ids': input_ids,
            'attention_mask': attention_mask,
            'examples': examples
        }



class LightningModule(pl.LightningModule):
    def __init__(self, hparams, data_module):
        super().__init__()
        self.save_hyperparameters(hparams)
        self.data_module = data_module
        
        if not self.hparams.no_generate:
            self.generator = T5ForConditionalGeneration.from_pretrained(self.hparams.generator_path)
            self.extractor = T5ForConditionalGeneration.from_pretrained(self.hparams.extractor_path)
    
            self.alignment_model = T5ForClassification.from_pretrained(self.hparams.alignment_model_path)
            self.fluency_model   = T5ForClassification.from_pretrained(self.hparams.fluency_model_path)

    def generate(self, batch):

        generated_ids = self.generator.generate(
             input_ids=batch['input_ids'], 
             attention_mask=batch['attention_mask'],
             num_return_sequences=1,
             max_length=self.hparams.max_seq_length,
             num_beams=1,
             do_sample=True,
             top_k=0,
             top_p=1,
        )
        generateds = self.decode(generated_ids)
        
        batch_encodings = self.tok(generateds)
        triplets_preds  = self.extractor.generate(
            input_ids=batch_encodings['input_ids'].cuda(),
            attention_mask=batch_encodings['attention_mask'].cuda(),
            num_return_sequences=1,
            max_length=self.hparams.max_seq_length,
            num_beams=1,
        )

        triplets_preds = self.decode(triplets_preds)
        alignment_scores, fluency_scores = self.cal_scores(generateds, batch)

        return {
            'generateds': generateds,
            'triplets_preds': triplets_preds,
            'alignment_scores': alignment_scores,
            'fluency_scores': fluency_scores,
            'examples': batch['examples'],
        }

    def cal_scores(self, generateds, batch):
        triplets_seqs = [example['triplets_seq'] for example in batch['examples']]
        triplets_generateds = [triplets_seq + ' ; ' + generated 
                               for triplets_seq, generated in zip(triplets_seqs, generateds)]

        batch_encodings = self.tok(triplets_generateds, max_length=140)
        alignment_scores = self.alignment_model(
            input_ids=batch_encodings['input_ids'].cuda(),
            attention_mask=batch_encodings['attention_mask'].cuda(),
        )
        
        batch_encodings = self.tok(generateds)
        fluency_scores = self.fluency_model(
            input_ids=batch_encodings['input_ids'].cuda(),
            attention_mask=batch_encodings['attention_mask'].cuda(),
        )
        return alignment_scores, fluency_scores

    def no_generate(self, batch):

        sentences = [example['triplet_sentence'] for example in batch['examples']]
        alignment_scores, fluency_scores = self.cal_scores(sentences, batch)

        return {
            'generateds': sentences,
            'triplets_preds': [''] * len(sentences),
            'alignment_scores': alignment_scores,
            'fluency_scores': fluency_scores,
            'examples': batch['examples'],
        }

    def predict_step(self, batch, batch_idx, dataloader_idx=0):
        if hasattr(self, 'generator'):
            self.generator.eval()

            self.extractor.eval()
            self.alignment_model.eval()
            self.fluency_model.eval()

            with torch.no_grad():
                if self.hparams.no_generate:
                    output = self.no_generate(batch)
                else:
                    output = self.generate(batch)

            output['dataloader_idx'] = dataloader_idx

            return output

    def decode(self, input_ids):
        return self.data_module.tokenizer.batch_decode(input_ids, skip_special_tokens=True)

    def tok(self, text, max_length=None):
        max_length = self.hparams.max_seq_length if max_length is None else max_length
        return tok(self.data_module.tokenizer, text, max_length)

    @staticmethod
    def add_model_specific_args(parser):
        parser.add_argument("--output_dir", type=str)
        parser.add_argument("--extractor_path", type=str)
        parser.add_argument("--alignment_model_path", type=str)
        parser.add_argument("--fluency_model_path", type=str)
        parser.add_argument("--no_generate", action='store_true')
        
        return parser



class CustomWriter(BasePredictionWriter):
    def __init__(self, output_dir, write_interval='epoch'):
        super().__init__(write_interval)
        self.output_dir = output_dir

    def write_on_epoch_end(self, trainer, pl_module, predictions, batch_indices):
        output_examples = []

        for output in tqdm(predictions[0]):

            examples = output['examples']
            generateds = output['generateds']
            triplets_preds = output['triplets_preds']
            alignment_scores = output['alignment_scores']
            fluency_scores = output['fluency_scores']

            for example, generated, triplets_pred, alignment_score, fluency_score in zip(examples, generateds, triplets_preds, alignment_scores, fluency_scores):

                example = {
                    'ID': len(output_examples),
                    'triplets_seqq': example['triplets_seq'],
                    'triplets_pred': triplets_pred,
                    'syn_seq': example['syn_seq'],
                    'generated_sentence': generated,
                    'triplet_sentence': example['triplet_sentence'],
                    'syn_sentence': example['syn_sentence'],
                    'alignment_score': float(alignment_score),
                    'fluency_score': float(fluency_score),
                }

                output_examples.append(example)

        output_file_name = os.path.join(self.output_dir, 'augmented_train.json')
        print(f'save {len(output_examples)} to', output_file_name)
        save_json(output_examples, output_file_name)



def main():
    parser = argparse.ArgumentParser()
    parser = pl.Trainer.add_argparse_args(parser)
    parser = LightningModule.add_model_specific_args(parser)
    parser = DataModule.add_argparse_args(parser)

    args = parser.parse_args()
    pl.seed_everything(args.seed)

    data_module = DataModule.from_argparse_args(args)
    data_module.load_dataset()

    model = LightningModule(args, data_module)

    pred_writer = CustomWriter(output_dir=args.output_dir)
    kwargs = {
        'callbacks': [pred_writer],
        'logger': False,
        'enable_checkpointing': False,
    }
    trainer = pl.Trainer.from_argparse_args(args, **kwargs)

    trainer.predict(model, datamodule=data_module, return_predictions=False)



if __name__ == '__main__':
    main()
 ```
根据参数output_dir： ./output/augmentation_filtered/14res_5k_42文件夹即为输出的结果，在我的环境上输出的结果如图所示。
![QQ截图20240328084211](https://github.com/HITSZ-HLT/T2S-Augmentation/assets/42504909/e373b748-60c3-4daf-8cbf-895110bf9ba4)
其中test.json为我想要的数据格式，但是train.json并不是我想要的格式，在此是否缺少一个转换代码将其转换回原始的数据格式？
感谢您的项目，它对我的研究工作提供了巨大的帮助！

**Comment by 1140310118** [comment]
你好，转换后的数据在 https://github.com/HITSZ-HLT/T2S-Augmentation/tree/main/data/augmented 中可以下载。
你可以调用utils.triplet中的parse_triplets_seq_tokens(triplets_seq, sentence, tokenizer)函数转换为你需要的格式。注意这里的tokenizer在创建的时候，需要设置use_fast=True，即

    tokenizer = AutoTokenizer.from_pretrained(model_name_or_path, use_fast=True)

**Comment by xlwreally** [comment]
您好，感谢您的项目，我已经成功的运行代码并获得了我想要的数据
